src/edu/csus/ecs/pc2/ui/ShadowControlPane.java

```diff
@@ -785,23 +977,51 @@ public void updateShadowNumberofRecords(int nRec)
      */
     public void connectFailed(String token)
     {
-        // TODO: Update JTable - future commit/PR
+        if(token == null || token.isEmpty()) {
+            addConnectTableEntry("Connection failed");
+        } else {
+            addConnectTableEntry("Connection failed starting at token " + token);
+        }
     }
 
     /**
      * {@inheritDoc}
      */
     public void connectSucceeded(String token)
     {
-        // TODO: Update JTable - future commit/PR
+        if(token == null || token.isEmpty()) {
+            addConnectTableEntry("Connected successfully");
+        } else {
+            addConnectTableEntry("Connected successfully starting at token " + token);
+        }
+    }
+    
+    /**
+     * {@inheritDoc}
+     */
+    public void connectClosed(String msg)
+    {
+        if(msg == null || msg.isEmpty()) {
+            msg = "Connection closed";
+        }
+        
+        // Save last token on disconnect
+        if(lastToken != null && !lastToken.isEmpty()) {
+            updateContestInformation();
+            msg += " at token " + lastToken;
+        }
+        addConnectTableEntry(msg);
     }
     
     /**
      * {@inheritDoc}
      */
     public void errorDisconnect(String errMsg)
     {
-        // TODO: Update JTable - future commit/PR
```

Comment:
What is this todo?   there should be some sort of summary description of the next task
or a reference to a github issue comment.

Reply:
@lane55 : I don't understand your comment on this line.  The line is a "deleted" line; it's an old "TODO" that has been "ToDONE" with the following few lines, just like the two previous "TODO" comments which have been removed and replaced by code implementing the "TODO"....